app/assets/javascripts/discourse/app/components/site-header.js

```diff
@@ -443,30 +443,41 @@ export default SiteHeaderComponent.extend({
 
   init() {
     this._super(...arguments);
-
     this._resizeObserver = null;
   },
 
+  updateHeaderOffset() {
+    // check offset from top of the page, in case there's a custom header above
+    // and adjust on scroll in case it's not sticky
+    if (this.scrollAnimationFrame) {
+      cancelAnimationFrame(this.scrollAnimationFrame);
+    }
+
+    this.scrollAnimationFrame = requestAnimationFrame(() => {
+      const currentHeaderWrapTop = this.headerWrap.getBoundingClientRect().top;
+
+      document.documentElement.style.setProperty(
+        "--header-offset",
+        `${this.headerWrap.offsetHeight + currentHeaderWrapTop}px`
+      );
+    });
+  },
+
   didInsertElement() {
     this._super(...arguments);
 
     this.appEvents.on("site-header:force-refresh", this, "queueRerender");
 
-    const headerWrap = document.querySelector(".d-header-wrap");
-    let header;
-    if (headerWrap) {
+    this.headerWrap = document.querySelector(".d-header-wrap");
+
+    if (this.headerWrap) {
       schedule("afterRender", () => {
-        header = headerWrap.querySelector("header.d-header");
-        const headerOffset = headerWrap.offsetHeight;
-        const headerTop = header.offsetTop;
-        document.documentElement.style.setProperty(
-          "--header-offset",
-          `${headerOffset}px`
-        );
-        document.documentElement.style.setProperty(
-          "--header-top",
-          `${headerTop}px`
-        );
+        this.header = this.headerWrap.querySelector("header.d-header");
+        this.updateHeaderOffset();
+      });
+
+      window.addEventListener("scroll", () => this.updateHeaderOffset(), {
+        passive: true,
```

Comment:
This potentially has a very large performance impact... ideally we should find some way to avoid using a scroll listener here (or at a minimum, avoid triggering style changes during scroll)